### PR TITLE
Shumin/map service

### PR DESCRIPTION
**Pull request by Shumin326**
* Changed to using our map service as default so the map can be updated periodically. 
* It has the feature that the **self-hold map variable will be updated only if received message is different**.
* If our map service is not launched, it will switch to using ROS map_server and get a static map. So if you still want to use the static map, launch ROS map_server along with the particle filter.

**TEST COVERAGE**: It now can get the initial map from our map_service and do localization well. However, when map is updated, node will throw error. It's due to that a lot of other variables dependent on the occupancy map need to be updated(not completely clear yet).

### CODE REVIEW

File: launch/localize_mapservice_viz.launch [hunk 1]
```diff
@@ -28,7 +28,7 @@
 		              publishes on /pf/pose/odom
 	 -->
 
-  <include file="$(find particle_filter)/launch/localize_nomapserver.launch"/>
+  <include file="$(find particle_filter)/launch/localize_mapservice.launch"/>
```
Comment (shineyruan) [inline launch/localize_mapservice_viz.launch:31]: Same problem as above

File: launch/localize_mapservice.launch [hunk 2]
```diff
@@ -27,6 +27,12 @@
 		publish_odom: whether or not to publish inferred pose as an odometry message.
 		              publishes on /pf/pose/odom
 	 -->
+	<!-- Launch the initial pose publisher node -->
+	<node pkg="particle_filter" type="initialpose_publisher.py" name="initialpose_publisher" output="screen"/>
+
+	<!-- Launch the customized map service node-->
+	<node pkg="map_service" type="revit_map_test" name="revit_map_test"/>
```
Comment (shineyruan) [inline launch/localize_mapservice.launch:34]: **Please avoid writing launch files that directly depends on `map_service` in this rospkg**, as this rospkg is meant to run onboard and `map_service` is meant to run on central server (handles multi-agent map merging as well in future). Normally they will be placed in different ROS workspaces. 

Writing a launch file that runs particle filter without any map is sufficient. We can manually run the map server in another terminal.

**Review by shineyruan — CHANGES_REQUESTED** [review]

**Comment by shineyruan** [comment]
Another quick question --- is the particle filter ready to subscribe `map_service`? It seems to me that this feature is not ready yet due to the crash on map updates, as reflected in your code comments.

**Comment by Shumin326** [comment]
> Another quick question --- is the particle filter ready to subscribe `map_service`? It seems to me that this feature is not ready yet due to the crash on map updates, as reflected in your code comments.

Fixed, thanks!

File: scripts/particle_filter.py [hunk 3]
```diff
@@ -175,16 +177,46 @@ def get_omap(self):
...
+      self.state_lock.acquire()
+      # TODO: It would be bettern to use map identifier instead
+      # of checking every entry of the map.
+      if self.map == None or (not np.array_equal(map_msg.data, self.map.data)):
+        rospy.loginfo("Map update succeeded!")
+        self.map = map_msg
+        self.MAX_RANGE_PX = int(self.MAX_RANGE_METERS / self.map.info.resolution)
+        self.init_range_method()
+        self.precompute_sensor_model()
+        self.state_lock.release()
```
Comment (shineyruan) [inline scripts/particle_filter.py:206]: Should not be inside the if statement

**Review by shineyruan — CHANGES_REQUESTED** [review]

File: scripts/particle_filter.py (hunk 3 above)
Comment (Shumin326) [reply scripts/particle_filter.py]: Good catch!

**Review by shineyruan — APPROVED** [review]
lgtm

**Pull request merged**